### PR TITLE
Add a PhoneNumber class to handle formatting and actions on numbers

### PR DESCRIPTION
**Pull request by caitlinrhd**
We value collaboration at Invoca, and focus on peer mentorship as a way for the whole team to learn and thrive. As part of this process, we conduct code reviews during our development cycle. In this phase of the interview, we would like you to review and comment on this Pull Request. You can assume that this Pull Request was written by an intern, who is new to professional software development.  Please take about 20-30 minutes to review this Pull Request and comment as if you were reviewing an intern's code.

### CODE REVIEW

File: phone_number.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  def format_number
+    "(#{find_area_code}) #{find_exchange}-#{find_line_number}"
+  end
+
+  def find_area_code
+    open_paren = @number.index('(')
+    close_paren = @number.index(')')
+    if open_paren
+      if close_paren
```
Comment (JQuinnie) [inline phone_number.rb:44]: Not a bad thought logic here, but what if the user didn't supply the closing paran, ")"? Would an area code get returned at all?

File: phone_number.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        @number[(open_paren + 1)..(close_paren - 1)]
+      end
+    else
+      @number.gsub!(/[^0-9]/, "")
+      @number[0..2]
```
Comment (JQuinnie) [inline phone_number.rb:49]: I think the patterns you have established for the other methods below will work for this as well, as long as you are expecting to receive 10 digits... so just keep this section.

File: phone_number.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(message, message)
+  end
+
+  def number
+    @number
+  end
+
+  def to_i
+    n = @number.gsub!(/[^0-9]/, "")
```
Comment (JQuinnie) [inline phone_number.rb:30]: I didn't see the method `to_i` used yet, was there an intention to use it?

Also this one is kind of tricky, be careful with `gsub!`, it will return `nil` if no substitutions were made, say on a phone number like "5558675309". See docs [here](https://ruby-doc.org/core-3.1.2/String.html)

File: phone_number.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
+require_relative "./phone_actions"
+
+class PhoneNumber
+  USERNAME = ENV['ACTIONS_USERNAME']
+  PASSWORD = ENV['ACTIONS_PASSWORD']
+
+  def initialize(number)
+    @number = number
+    @number = format_number
```
Comment (JQuinnie) [inline phone_number.rb:9]: this could be more of a nit/preference, `number` is used a lot and could be kind of confusing under different contexts, I would probably remove this line, L9 and update the method on L25 below with something like:

```
def formatted_number
  @number = format_number
end
```

as that can give a clear idea of my intentions that numbers will be formatted from that method.

File: phone_actions.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
+
+  def initialize(username, password)
+    @actions_url = "https://#{username}:#{password}@api.<made_up>.com/actions/"
+  end
+
+  def call!(number, message)
+    JSON.parse(RestClient.post("@actions_url/#{text}", {message: message, number: number}))
```
Comment (JQuinnie) [inline phone_actions.rb:8]: Did you mean to include the variable in the string literal so that `RestClient` will receive the full URL?

```suggestion
    JSON.parse(RestClient.post("${@actions_url}/#{text}", {message: message, number: number}))
```

Comment (JQuinnie) [reply phone_actions.rb:8]: For both this line and L12

File: phone_actions.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
+
+  def initialize(username, password)
+    @actions_url = "https://#{username}:#{password}@api.<made_up>.com/actions/"
+  end
+
+  def call!(number, message)
```
Comment (JQuinnie) [inline phone_actions.rb:7]: I would probably not use the `!` here for on the `text!()` below, as there isn't a complementary method that is 'safer'... also, this method is not an in-place, destructive method, nor does it mutate.

File: phone_actions.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
```
Comment (JQuinnie) [inline phone_actions.rb:1]: Nice job setting these methods up to use `RestClient`, one thing I would try to be aware of is what are your expectations for error handling? What if the call to `RestClient` fails? Do you want raise exceptions and manage the error messages?

Take a look at the [rest-client docs](https://github.com/rest-client/rest-client) under 'Result handling'

File: phone_number_spec.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
+require_relative "./phone_number"
```
Comment (JQuinnie) [inline phone_number_spec.rb:1]: Great specs so far!

Don't be afraid to include more error and edge cases, additional phone numbers to test.

File: phone_number_spec.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+RSpec.describe PhoneNumber do
+  let(:phone_number) { PhoneNumber.new("5558675309") }
+
+  describe "#number" do
+    it "is a formatted number" do
+      expect(phone_number.number).to eq("(555) 867-5309")
+    end
+  end
+  
+  describe "#to_i" do
```
Comment (JQuinnie) [inline phone_number_spec.rb:12]: When you include a new phone number case like `"5558675309"` it might produce something different than what you expect.

**Review by JQuinnie — COMMENTED** [review]
Amazing work on this so far! ⭐️

I provided some feedback on mostly edge cases and the unexpected, happy to discuss any of the comments in more detail.

Keep up the good work!
